modules/3d/src/rgbd/odometry_frame_impl.cpp

```diff
@@ -0,0 +1,250 @@
+        OdometryFramePyramidType pyrType, size_t level) override;
+
+private:
+    void findMask(InputArray image);
+
+    TMat image;
+    TMat imageGray;
+    TMat depth;
+    TMat mask;
+    TMat normals;
+    std::vector< std::vector<TMat> > pyramids;
+};
+
+OdometryFrame::OdometryFrame()
+{
+    this->impl = makePtr<OdometryFrameImplTMat<Mat>>();
+};
+
+OdometryFrame::OdometryFrame(OdometryFrameStoreType matType)
+{
+    if (matType == OdometryFrameStoreType::UMAT)
+        this->impl = makePtr<OdometryFrameImplTMat<UMat>>();
+    else
+        this->impl = makePtr<OdometryFrameImplTMat<Mat>>();
+};
+
+void OdometryFrame::setImage(InputArray  image) { this->impl->setImage(image); }
+void OdometryFrame::getImage(OutputArray image) const { this->impl->getImage(image); }
+void OdometryFrame::getGrayImage(OutputArray image) const { this->impl->getGrayImage(image); }
+void OdometryFrame::setDepth(InputArray  depth) { this->impl->setDepth(depth); }
+void OdometryFrame::getDepth(OutputArray depth) const { this->impl->getDepth(depth); }
+void OdometryFrame::setMask(InputArray  mask) { this->impl->setMask(mask); }
+void OdometryFrame::getMask(OutputArray mask) const { this->impl->getMask(mask); }
+void OdometryFrame::setNormals(InputArray  normals) { this->impl->setNormals(normals); }
+void OdometryFrame::getNormals(OutputArray normals) const { this->impl->getNormals(normals); }
+void OdometryFrame::setPyramidLevel(size_t _nLevels, OdometryFramePyramidType oftype)
+{
+    this->impl->setPyramidLevel(_nLevels, oftype);
+}
+void OdometryFrame::setPyramidLevels(size_t _nLevels) { this->impl->setPyramidLevels(_nLevels); }
+size_t OdometryFrame::getPyramidLevels(OdometryFramePyramidType oftype) const { return this->impl->getPyramidLevels(oftype); }
+void OdometryFrame::setPyramidAt(InputArray  img, OdometryFramePyramidType pyrType, size_t level)
+{
+    this->impl->setPyramidAt(img, pyrType, level);
+}
+void OdometryFrame::getPyramidAt(OutputArray img, OdometryFramePyramidType pyrType, size_t level) const
+{
+    this->impl->getPyramidAt(img, pyrType, level);
+}
+
+template<typename TMat>
+OdometryFrameImplTMat<TMat>::OdometryFrameImplTMat()
+    : pyramids(OdometryFramePyramidType::N_PYRAMIDS)
+{
+};
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::setImage(InputArray _image)
+{
+    this->image = getTMat<TMat>(_image);
+    Mat gray;
+    if (_image.channels() != 1)
+        cvtColor(_image, gray, COLOR_BGR2GRAY, 1);
+    else
+        gray = getTMat<Mat>(_image);
+    gray.convertTo(gray, CV_8UC1);
+    this->imageGray = getTMat<TMat>(gray);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::getImage(OutputArray _image)
+{
+    _image.assign(this->image);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::getGrayImage(OutputArray _image)
+{
+    _image.assign(this->imageGray);
+}
+
+template<typename TMat>
+void OdometryFrameImplTMat<TMat>::setDepth(InputArray _depth)
```

Comment:
1. I guess you used `Mat` as a temporary storage for depth preprocessing. You'd better use TMat for that and call `getTMat<TMat>(_depth)` instead of `getMat()`.
2. A comment explaining these magic numbers should be added about the workaround used here, like "ICP fails on depth values bigger than NNNN, let's keep scaled down version".

Reply:
Also, use TMat instead of Mat in this function